### ISSUE TITLE
`generate_text_basic` function changes model mode as a side effect (ch02_main.ipynb)

### ISSUE THREAD

**Issue by labdmitriy** [rasbt/reasoning-from-scratch#118]
Hi Sebastian,

`generate_text_basic` function uses `model.eval()` that changes the model mode if it was in training mode before.
This is also in our case:
```python
model.training
# True
...
output_token_ids_tensor = generate_text_basic(
    model=model,
    token_ids=input_token_ids_tensor,
    max_new_tokens=max_new_tokens,
)
...
model.training
# False
```

We can use context manager to save original mode and restore it after the inference:
```python
from contextlib import contextmanager

@contextmanager
def eval_mode(model):
    original_mode = model.training
    model.eval()
    
    try:
        yield
    finally:
        model.train(original_mode)

@torch.inference_mode()
def generate_text_basic(
    model,
    token_ids,
    max_new_tokens,
    eos_token_id=None
):
    input_length = token_ids.shape[1]
    
    with eval_mode(model):
        for _ in range(max_new_tokens):
            out = model(token_ids)[:, -1]
            next_token = torch.argmax(out, dim=-1, keepdim=True)
    
            # Stop if all sequences in the batch have generated EOS
            if (eos_token_id is not None and next_token.item() == eos_token_id):
                break
    
            token_ids = torch.cat([token_ids, next_token], dim=1)
    return token_ids[:, input_length:]
```
Then this function will not have such side effect:
```python
model.training
# True
...
output_token_ids_tensor = generate_text_basic(
    model=model,
    token_ids=input_token_ids_tensor,
    max_new_tokens=max_new_tokens,
)
...
model.training
# True
```

**Comment by rasbt** [comment]
Thanks! I think this is an elegant solution. Unfortunately, related to the other PR, I worry that this adds too much complexity to the chapter. I understand the main point to keep things side-effect free, but since we are not doing any training, this would be kind of overkill. 

Side note: The other challenge is that each code line can only have 76 characters (because otherwise it wraps the line in the Manning layout), so context managers can be a bit tricky, which is why I use the decorators instead of the context managers when it comes to `torch.no_grad()` and `torch.inference_mode()`.

Back to the topic of eval and training mode. In chapter 6 I am tracking that with a simple variable to override the side effects. I suppose we could also do this in this function here. It's leaner than the context manager (but does the same thing directly). Of course it's less cool but maybe more readable to a beginner?

```python
                if eval_on_checkpoint and math500_eval_data:
                    was_training = model.training
                    model.eval()
                    subset = (
                        math500_eval_data[:eval_max_items]
                        if eval_max_items
                        else math500_eval_data
                    )
                    out_path = (
                        Path(checkpoint_dir)
                        / f"{SCRIPT_NAME}-step{current_step:05d}-math500.jsonl"
                    )
                    num_correct, num_examples, acc = evaluate_math500_stream(
                        model=model,
                        tokenizer=tokenizer,
                        device=device,
                        math_data=subset,
                        out_path=out_path,
                        max_new_tokens=max_new_tokens,
                        verbose=False,
                    )
                    append_eval_metrics(current_step, acc, num_correct, num_examples)
                    print(
                        f"MATH-500 eval @ step {current_step}: "
                        f"acc={acc:.3f} ({num_correct}/{num_examples})"
                    )
                    if was_training:
                        model.train()
```

(Via [rlvr_grpo_original_no_kl.py](https://github.com/rasbt/reasoning-from-scratch/blob/main/ch06/02_rlvr_grpo_scripts_original/rlvr_grpo_original_no_kl.py#L244))


This is nice code though; let me keep this open to see if I can use it in bonus materials later.

**Comment by labdmitriy** [comment]
Hi Sebastian,

Yes I think this also possible solution and I suppose that it is better than having side effect because we work in Jupyter notebook and user can run different cells in different order and it can be not clear where and why the model mode was changed somewhere.

Also maybe we can combine both solutions and do more general save/restore like the following:
```python
if eval_on_checkpoint and math500_eval_data:
    was_training = model.training
    model.eval()
    ...
    model.train(was_training)
```

**Comment by rasbt** [comment]
Just submitted chapter 7 and am now slowly catching up with some of the other things... 

I don't mind simplifying 

```python
if was_training:
    model.train()
```

to 

```python
model.train(was_training)
```

But I do think the implicit `if` statement is a bit easier to read. It's more like human language "if the model was training, then put it back in training mode" whereas for `model.train(was_training)` I have to look a bit more carefully. Ie. it took me a couple of sec to see that this indeed also works, which tells me that it may be a tad less intuitive (but it may just be me). It's a good and elegant suggestion though.

**Comment by rasbt** [comment]
I was playing with different ideas here, and turning the eval-mode toggling into a decorator would also be a neat way to avoid the extra indentation that the contextlib brings:

```python
import functools
import torch


def eval_mode(fn):
    @functools.wraps(fn)
    def wrapper(model, *args, **kwargs):
        original_mode = model.training
        model.eval()
        try:
            return fn(model, *args, **kwargs)
        finally:
            model.train(original_mode)
    return wrapper


@eval_mode
@torch.inference_mode()
def generate_text_basic(
    model,
    token_ids,
    max_new_tokens,
    eos_token_id=None
):
    input_length = token_ids.shape[1]
    # ...
```

I am tempted to add this, but I do think though that this may look a bit overwhelming for PyTorch beginners. I know from the previous book that a lot of readers learn PyTorch (and sometimes Python) along the way, and this would just add more code and complexity.

We are not using any training until chapter 6 anyways. And while this keeps the side effect, `model.eval()` is just a simpler approach here.

I was also tempted to remove it, and I actually had a PR ready for that, but then I thought that since we don't do any training until later, having that `model.eval()` in there doesn't really hurt. But then it's probably good to also add an explicit `model.train()` in the training function, which I will do via #172
